pkg/nsx/services/builder.go

```diff
@@ -294,21 +307,46 @@ func (service *SecurityPolicyService) buildExpressionsMatchExpression(matchExpre
 	return err
 }
 
-func (service *SecurityPolicyService) buildPolicyGroupID(obj *v1alpha1.SecurityPolicy) string {
+// build appliedTo group ID for both policy and rule levels.
+func (service *SecurityPolicyService) buildAppliedGroupID(obj *v1alpha1.SecurityPolicy, ruleIdx int) string {
+	if ruleIdx != -1 {
+		return fmt.Sprintf("sp_%s_%d_scope", obj.UID, ruleIdx)
+	}
+
 	return fmt.Sprintf("sp_%s_scope", obj.UID)
 }
 
-func (service *SecurityPolicyService) buildPolicyGroupPath(obj *v1alpha1.SecurityPolicy) string {
-	policyGroupID := service.buildPolicyGroupID(obj)
-	return fmt.Sprintf("/infra/domains/%s/groups/%s", getDomain(service), policyGroupID)
+// build appliedTo group display name for both policy and rule levels.
+func (service *SecurityPolicyService) buildAppliedGroupName(obj *v1alpha1.SecurityPolicy, ruleIdx int) string {
+	var rule *v1alpha1.SecurityPolicyRule
+	if ruleIdx != -1 {
+		rule = &(obj.Spec.Rules[ruleIdx])
+		if len(rule.Name) > 0 {
+			return fmt.Sprintf("%s-scope", rule.Name)
+		}
+		return fmt.Sprintf("%s-%d-scope", obj.ObjectMeta.Name, ruleIdx)
+	}
+
+	return fmt.Sprintf("%s-%s-scope", obj.ObjectMeta.Namespace, obj.ObjectMeta.Name)
+}
+
+// build appliedTo group path for both policy and rule levels.
+func (service *SecurityPolicyService) buildAppliedGroupPath(obj *v1alpha1.SecurityPolicy, ruleIdx int) string {
+	var groupID string
+	if ruleIdx == -1 {
+		groupID = service.buildAppliedGroupID(obj, -1)
+	} else {
+		groupID = service.buildAppliedGroupID(obj, ruleIdx)
+	}
```

Comment:
It looks that L335-L340 can be simplified to:
`groupID = service.buildAppliedGroupID(obj, ruleIdx)`

Reply:
Thanks, good point. I've changed the code as advised.